src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityReplyChannelBinder.cs

```diff
@@ -0,0 +1,201 @@
+
+            if (channel == null)
+            {
+                Fx.Assert("ReplyChannelBinder.ReplyChannelBinder: (channel != null)");
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(channel));
+            }
+            _channel = channel;
+            ListenUri = listenUri;
+            _initialized = true;
+            if(_channel is SecurityReplyChannel)
+            {
+                this._securityDispatcher = ((SecurityReplyChannel)_channel).SecurityServiceDispatcher;
+            }
+        }
+
+        public IChannel Channel
+        {
+            get { return _channel; }
+        }
+
+        public bool HasSession
+        {
+            get { return _channel is SecurityReplyChannel; }
+        }
+
+        public Uri ListenUri { get; private set; }
+
+        public EndpointAddress LocalAddress
+        {
+            get { return _channel.LocalAddress; }
+        }
+
+        public EndpointAddress RemoteAddress
+        {
+            get
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotImplementedException());
+            }
+        }
+
+        public void Abort()
+        {
+            _channel.Abort();
+        }
+
+        public void CloseAfterFault(TimeSpan timeout)
+        {
+            var helper = new TimeoutHelper(timeout);
+            _channel.CloseAsync(helper.GetCancellationToken()).GetAwaiter().GetResult();
+        }
+
+        public RequestContext CreateRequestContext(Message message)
+        {
+            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotImplementedException());
+        }
+
+        public Task SendAsync(Message message, CancellationToken token)
+        {
+            throw TraceUtility.ThrowHelperError(new NotImplementedException(), message);
+        }
+
+        public Task<Message> RequestAsync(Message message, CancellationToken token)
+        {
+            throw TraceUtility.ThrowHelperError(new NotImplementedException(), message);
+        }
+
+        public void SetNextDispatcher(IServiceChannelDispatcher dispatcher)
+        {
+            _next = dispatcher;
+        }
+
+        public Task DispatchAsync(RequestContext context)
+        {
+            Fx.Assert(_next != null, "SetNextDispatcher wasn't called");
+            if (IdentityModel.SecurityUtils.
+                IsRequestSecurityContextIssuance(context.RequestMessage.Headers.Action))
+            {
+                context = HandleHandshake(context);
+                //Add logic to separate flow for handshake or actual call
+                return _next.DispatchAsync(context);
+            }
+            else
+            {
+                return ProcessRequest(context);
+
+            }
+
+        }
+
+        public Task DispatchAsync(Message message)
+        {
+            throw new NotImplementedException();
+        }
+
+
+
+
+
+        private Task ProcessRequest(RequestContext context)
+        {
```

Comment:
Can you add a comment to explain where these method originated. For example, this one looks to have come from SecuritySessionServerSettings.SessionInitiationMessageHandler.
Also rename to ProcessRequestAsync

Reply:
TODO : Delete